container/apbf/filter.go

```diff
@@ -0,0 +1,624 @@
+	// the logical slice was just rotated once backwards around the ring buffer
+	// above, this clears what was previously the oldest slice so the new
+	// entries take its place.
+	startBit := uint64(f.baseIndex) * f.bitsPerSlice
+	endBit := startBit + f.bitsPerSlice
+
+	// Clear bits up to the next byte boundary.
+	byteIdx := startBit >> 3
+	if bit := startBit & 7; bit != 0 {
+		f.data[byteIdx] &= byte(1<<bit - 1)
+		byteIdx++
+	}
+
+	// Clear full bytes in one fell swoop when possible.
+	if endByteIdx := (endBit >> 3); endByteIdx > byteIdx {
+		fullBytes := endByteIdx - byteIdx
+		data := f.data[byteIdx : byteIdx+fullBytes]
+		for i := range data {
+			data[i] = 0
+		}
+		byteIdx += fullBytes
+	}
+
+	// Clear any remaining bits.
+	if byteIdx < uint64(len(f.data)) {
+		f.data[byteIdx] &^= byte(1<<(endBit&7) - 1)
+	}
+
+	f.itemsInCurGeneration = 0
+}
+
+// deriveIndex uses enhanced double hashing to calculate a unique index for the
+// given logical slice number using the closed formula.  In order to support
+// more efficient interleaving of the calculation while looping over several
+// slices, it also returns the intermediate accumulator which can be fed into
+// subsequent iterations of the algorithm to potentially provide significant
+// speed improvements.
+//
+// Note that each slice conceptually uses an independent hash function to
+// determine which bit to index.  However, rather than using the more
+// traditional approach of a distinct hash function for each one, enhanced
+// double hashing is used to effectively generate the necessary hash function
+// because it is significantly faster while still coming quite close to the
+// theoretical limit of two-index fingerprinting.
+//
+// It is defined as "f(i) = hash1 + i*hash2 + (i^3 - i)/6 (mod m)", where m is
+// the number of bits to index, however, this function does not perform the
+// modular reduction and instead leaves it up to the caller.
+//
+// It is also worth noting that enhanced double hash is used in this
+// implementation over the more common double hashing referenced in [APBF], and
+// defined in [LHSP], which generates the necessary hash functions from a linear
+// combination of the two independent hash functions as "f(i) = hash1 + i*hash2
+// (mod m)" because, as pointed out in [BFPV], that construction results in
+// imposing an observable accuracy limit since there are no restrictions placed
+// on the second hash to ensure the resulting indices are (almost) all unique.
+// Namely, when the second hash function produces 0, all indices are the same,
+// and, similarly, when it produces a value that divides m, the number of
+// indices probed will be the minimum of the number of slices and the produces
+// value, which could be as small as 2.
+//
+// See section 5.2 of [BFPV] for further details, if desired.
+func deriveIndex(slice uint16, hash1, hash2 uint64) (uint64, uint64) {
+	// hash1 + slice*hash2 + ((slice^3 - slice) / 6)
```

Comment:
Do you intend to remove this?

Reply:
I had left it there intentionally, but fair point that it's up in the comment for the function already, so not needed.

Removed.